lib/api/CPerPartitionCategoryIdMapper.cc

```diff
@@ -0,0 +1,95 @@
+                                                     model::CLocalCategoryId localCategoryId) {
+    if (localCategoryId.isValid() == false) {
+        return CGlobalCategoryId{localCategoryId.id()};
+    }
+    auto iter = m_Mapper.find(categorizerKey);
+    if (iter == m_Mapper.end()) {
+        iter = m_Mapper.emplace(categorizerKey, TGlobalCategoryIdVec{}).first;
+    }
+    TGlobalCategoryIdVec& partitionVec{iter->second};
+    std::size_t index{localCategoryId.index()};
+    if (index > partitionVec.size()) {
+        LOG_ERROR(<< "Bad category mappings: " << (index - partitionVec.size())
+                  << " local to global category ID mappings missing for partition "
+                  << categorizerKey);
+        partitionVec.resize(index);
+    }
+    if (index == partitionVec.size()) {
+        partitionVec.emplace_back(++m_HighestGlobalId, iter->first, localCategoryId);
+    }
+    return partitionVec[index];
+}
+
+CCategoryIdMapper::TCategoryIdMapperUPtr CPerPartitionCategoryIdMapper::clone() const {
+    return std::make_unique<CPerPartitionCategoryIdMapper>(*this);
+}
+
+void CPerPartitionCategoryIdMapper::acceptPersistInserter(core::CStatePersistInserter& inserter) const {
+
+    inserter.insertValue(HIGHEST_GLOBAL_ID_TAG, m_HighestGlobalId);
+
+    for (const auto& mapEntry : m_Mapper) {
+        inserter.insertValue(CATEGORIZER_KEY_TAG, mapEntry.first);
+        for (const auto& globalCategoryId : mapEntry.second) {
+            inserter.insertValue(GLOBAL_ID_TAG, globalCategoryId.globalId());
+        }
+    }
+}
+
+bool CPerPartitionCategoryIdMapper::acceptRestoreTraverser(core::CStateRestoreTraverser& traverser) {
+
+    m_HighestGlobalId = 0;
+    m_Mapper.clear();
+
+    auto iter = m_Mapper.end();
+    do {
+        const std::string& name{traverser.name()};
+        if (name == HIGHEST_GLOBAL_ID_TAG) {
+            if (core::CStringUtils::stringToType(traverser.value(), m_HighestGlobalId) == false) {
+                LOG_ERROR(<< "Invalid highest global ID in " << traverser.value());
+                return false;
+            }
+        } else if (name == CATEGORIZER_KEY_TAG) {
+            iter = m_Mapper.emplace(traverser.value(), TGlobalCategoryIdVec{}).first;
+        } else if (name == GLOBAL_ID_TAG) {
+            if (iter == m_Mapper.end()) {
+                LOG_ERROR(<< "Global ID seen before categorizer key "
+                          << traverser.value());
+                return false;
+            }
+            int globalId{model::CLocalCategoryId::SOFT_CATEGORIZATION_FAILURE_ERROR};
+            if (core::CStringUtils::stringToType(traverser.value(), globalId) == false) {
+                LOG_ERROR(<< "Invalid global ID in " << traverser.value());
+                return false;
+            }
+            iter->second.emplace_back(globalId, iter->first,
+                                      model::CLocalCategoryId{iter->second.size()});
+        }
```

Comment:
My inclination would be to have persist and restore functions on `CGlobalCategoryId` then you should be able to do `core::CPersistUtils::restore(MAPPER_TAG, m_Mapper, traverser)` and similar in `acceptPersistInserter`, but I'll leave it up to you whether you feel this is worthwhile.

Reply:
If persistence was at the level of `CGlobalCategoryId` and was encapsulated it would have to persist the categorizer key and local category ID.  By having it at this level it can get away with persisting the categorizer key just once per partition, and not persisting the local ID at all (because it can be derived from the vector index).  I am not convinced it would be helpful to have persistence methods on `CGlobalCategoryId` that weren't encapsulated, i.e. persistence only partially persisting the state and restore relying on the caller passing in significant other information that wasn't in the persisted state.

Reply:
Ok that makes sense.